### DISCUSSION TITLE
Fixing Ares' tile placement issues

### DISCUSSION THREAD

**Discussion by kberg** [bafolts/terraforming-mars#3514] [General]
A fairly regular complaint about Ares is reports of bugs like #1823, a situation where the server thinks a player can place tiles, but the associated costs with tile placement under certain circumstances makes it impossible to do. It turns out the few places the server thinks are valid are not, because they either require some uncalcuated megacredit costs (reds, Hellas, placing on top of hazards) or some uncalculated production costs (hazard adjacency.)

Anyone's first instinct is to just make the server-side calculations better -- if a placement costs extra megacredits, just check, while calculating tile placement, whether the player can afford the additional costs. The thing is, it's not quite that simple, for instance, placing next to two hazards, et cetera.

I made two attempts to start to address this in the past: #1842 and #1549, both of which didn't get too far.

One of the alternatives I had suggested was being able to completely cancel an action if it isn't playing out the way you like. The biggest concern about this is (and undo in general) is that a player can draw cards, and then choose to undo. There is plenty of discussion about the risks of a bad undo: #1846, for example.

One suggestion was to delay drawing cards until the end of the action, but in #1549 @Lynesth suggested a case where drawing cards earlier rather than later could be desirable.

I'll stop here, and then start talking about steps to address this.

**Comment by kberg** [comment]
First, I'll recap pieces of a discussion pasted in #1549 -- see there for the full conversation:

> The problem comes down to being one of defining what the app should do in best effort. So I see two options: 1. Be perfectly predictive. 2. Be reasonably predictive, but permissive, and also forgiving.

I hope above I've made clear how difficult it is to be perfectly predictive. So how can the app become more forgiving, while respecting the user interface clicks, drawn cards, and so on?

To do this I suggest four to six changes.
1. Game saves not once per turn, but once per action. The reason for this one will be clear later. But this does mean that games will now take up almost twice as many saves. Almost. Not quite, but close enough to not debate it here.
2. When placing a tile, add a button "Do not place a tile." A player who played a project that places two oceans might be half way through placing oceans when the second placement becomes too expensive to place, but too complicated to calculate whether it may be placed. A player at this point either must get stuck, or barrel forward to just realize they can't play the tile.
3. Provide a "Reset this action" button. The action would perform effectively a one-step undo, or, a zero-step undo, or, a reload.
4. Disallow "Reset this action" if the player has drawn a card this actrion.
5. (not required) add a confirm action before a player commits their final change.
6. Make sure cards are not drawn until a tile is fully paid for (probably already done, actually.)

There are some interesting outliers that I'm not going to get into (eg playing two oceans, wanting to cancel after the first ocean was played) but I think this is enough.

Some of these behaviors are of the kind people would want optional or disabled. I don't blame them.

**Comment by bafolts** [comment]
I don't think this necessarily needs to be one or the other. The server side change to enforce this on the server seems needed. The improvements for undo action also seem needed. I do acknowledge that calculating if a tile can be placed will not be straight forward.

For suggestion 1 of games saving once per action I think this will be fine. I think often times we worry about impact to storage before it is a problem. Anyone running the server locally it shouldn't be a problem. If it became a problem for popular sites we could look into optimizing storage more. We have made improvements to storage but we are by no means close to how small we could get it.

On the server side creating some mechanism where there are actions a player has taken that other players can not see we could store in some queue. When a user then finally clicked to confirm we would process all the items in the queue.

If I were starting over I would try to implement most the entire game logic as steps that can roll back. For example, when some action was taken which increments a player's mega-credits there would game a queue with some logic that increments the player's mega-credits. At any point then we could pop items off of this queue to go back in time. This should be similar to how git works where git only stores diffs as you go. As you rollback you start popping off diffs. Right now we take a snapshot of the entire game and store that in the database. If we had this queue we could potentially only store the diffs as entries versus the entire game state at any given moment.

**Reply by kberg** [reply]
I've thought about that inherent undo-type mechanism. It's that standard Command design pattern we learned about years ago. Clever. But, could you imagine implementing that now? That seems like a herculean effort now. Not sure it's worth taking on. Not without a really well-defined pattern to accompany it.

I'll start looking at more reliable save patterns, although I'm a little nervous about finding the right spot to save -- it's not really clear where to do something like that in this game. I wish the state machine part of `Player.ts` was more clear before taking that step on.

**Comment by kberg** [comment]
I think this is my next big effort. I don't think this is going to be "big" in the same way some of the other efforts are, but it'll be a good change, I think.

**Comment by kberg** [comment]
Update 2 years later: Adjacency is very significantly improved because processing eligible board spaces informs the caller about expected costs. Going to mark this as closed.
